### PR TITLE
Add keyword generics initiative announcement post to the internals blog

### PR DESCRIPTION
**Pull request by yoshuawuyts**
This adds the announcement of the keyword generics initiative to the rust internals blog. I've set the release date for Monday the 25th of July, and marked the blog post as `release: false` in case it gets merged before then.

Thanks!

cc/ @rust-lang/initiative-keyword-generics

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
The release keyword is a marker for Rust releases, used to produce https://blog.rust-lang.org/releases.json (I believe currently feeding the version number on https://www.rust-lang.org/, but I'm not sure).

So it won't help with accidental merging :)

**Comment by Mark-Simulacrum** [comment]
It also looks like this post is in posts/, not posts/inside-rust/, which is presumably an oversight? It sounds like you intended to put this on Inside Rust.

File: posts/2022-07-25-keyword-generics.md
Comment (Mark-Simulacrum) [inline posts/2022-07-25-keyword-generics.md:35]:
```suggestion
be generic over the qualifier keywords of functions and traits. The post ["What
```

File: posts/2022-07-25-keyword-generics.md
Comment (Mark-Simulacrum) [inline posts/2022-07-25-keyword-generics.md:130]: I feel like this example is missing something, since to my knowledge `42 + 1` worked for a long time already in const contexts (e.g., `const F: u32 = 42 + 1;`). The only time you might need the complex invocation is if you're using this as a const generic, but there a `const fn` isn't really going to help you either.

Maybe I'm the one missing something though :)

File: posts/2022-07-25-keyword-generics.md
Comment (Mark-Simulacrum) [inline posts/2022-07-25-keyword-generics.md:156]: I would avoid "we" here, since this is coming as a post from/on behalf of the keywords generics initiative. AFAIK, the initiative isn't intending to actively maintain or anything like that async-std, just make use of the valuable lessons learned in its design and usage -- I think this formulation mixes that up a bit too much. (Particularly later in the paragraph, around being a proving ground and successful proof).

File: posts/2022-07-25-keyword-generics.md
Comment (Mark-Simulacrum) [inline posts/2022-07-25-keyword-generics.md:185]: The problem I recall is that you get panics if you use a reqwest::blocking API from within an existing tokio runtime (called block_on in async context...) or something like that. Maybe with `futures::...` that's not a problem. Maybe keyword generics help here, though.

Mainly commenting since I think finding and using the right crates seems like.. well, a problem we're not going to solve with keyword generics, so it feels a little odd to comment on it. Crate authors may be able to more easily expose sync and async APIs, though; that seems mostly true. The problem I note with blocking interfaces being wrong to use in an async context (even if not in an async fn), though, doesn't seem like one you can tackle unless you have runtime dispatch (dyn keywords?). And the blocking is presumably mostly a problem only when you're blocking in an async context; if you're intending to block the thread my guess is the overhead is not that high?

File: posts/2022-07-25-keyword-generics.md
Comment (Mark-Simulacrum) [inline posts/2022-07-25-keyword-generics.md:255]:
```suggestion
"effects" will face the same issue. "fallibility" is particularly on our mind here,
```

File: posts/2022-07-25-keyword-generics.md
Comment (Mark-Simulacrum) [inline posts/2022-07-25-keyword-generics.md:280]: I feel like this isn't quite right -- presumably, the user still needs to choose what kind of I/O model they're using, since that has major performance (at least) implications, and plausibly correctness ones around timeouts etc.

I think the right claim earlier would be that the "end/final/binary" user must choose, but perhaps the individual developer of a library (e.g., of serde) may not need to choose. That resonates more with me.

File: posts/2022-07-25-keyword-generics.md
Comment (Mark-Simulacrum) [inline posts/2022-07-25-keyword-generics.md:296]:
```suggestion
### Q: Are you looking at other keywords beyond `async` and `const`?
```

File: posts/2022-07-25-keyword-generics.md
Comment (Mark-Simulacrum) [inline posts/2022-07-25-keyword-generics.md:24]: I think it's probably worth explicitly noting somewhere in this post that initiatives are really intended as explore/design space - there's not actually any guarantees or even strong confidence at this early stage in the process that any of the particulars will be adopted by Rust. Many of our initiatives come to fruition in somewhat different forms or not at all, particularly those relatively early in the process.

Might make for a good FAQ question (something like "Q: Where is the RFC?", perhaps?).

File: posts/2022-07-25-keyword-generics.md
Comment (fee1-dead) [inline posts/2022-07-25-keyword-generics.md:322]:
```suggestion
have some ideas of how we're going to make this work though, and we're
```

Saw this on the hackmd, then saw this PR :)

File: posts/2022-07-25-keyword-generics.md
Comment (fee1-dead) [inline posts/2022-07-25-keyword-generics.md:339]:
```suggestion
[fee1-dead][fd], [Daniel Henry-Mantilla][dhm], and [Ryan Levick][rylev]_
```

Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md:338]: Ohh, good catch. Thank you!

File: posts/2022-07-25-keyword-generics.md
Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md:279]: > I think the right claim earlier would be that the "end/final/binary" user must choose, but perhaps the individual developer of a library (e.g., of serde) may not need to choose. That resonates more with me.

Yeah, that's exactly what we're going for as well. I agree that the current phrasing doesn't quite capture that. Let me find a way to update that.

File: posts/2022-07-25-keyword-generics.md
Comment (rylev) [inline posts/2022-07-25-keyword-generics.md:33]: This paragraph seems to jump from "being able to be generic over types" to "being generic over asyncness" which is a bit confusing. It seems like you're describing what this feature *isn't* then an *example* of what this feature i. Yet you never attempt to define what the feature itself is. 

It would be nice to say something to the effect of "Rust allows you to be generic over types - it does not allow you to be generic over other things that are usually specified by keywords in Rust. For example, whether a function is async, whether a function can fail or not, whether a function is const or not, etc."

File: posts/2022-07-25-keyword-generics.md
Comment (rylev) [inline posts/2022-07-25-keyword-generics.md:51]: I would explicitly note that this syntax is just for demonstration is not really the point. The point is what it allows one to express.

File: posts/2022-07-25-keyword-generics.md
Comment (rylev) [inline posts/2022-07-25-keyword-generics.md:85]: You mention fallibility but you did not mention before that this is something that this feature can make generic.

File: posts/2022-07-25-keyword-generics.md
Comment (rylev) [inline posts/2022-07-25-keyword-generics.md:90]: This line makes this effort feel very specific to asyncness but it's my understanding this is meant to be much more general.

File: posts/2022-07-25-keyword-generics.md
Comment (rylev) [inline posts/2022-07-25-keyword-generics.md:171]: These comments are confusing. I suppose they're meant to indicate the modules these would live in, but that is very much not clear.

File: posts/2022-07-25-keyword-generics.md
Comment (rylev) [inline posts/2022-07-25-keyword-generics.md:5]:
```suggestion
```

false is the default

File: posts/2022-07-25-keyword-generics.md
Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md:23]: Good call; done!

File: posts/2022-07-25-keyword-generics.md
Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md:155]: Yep, that makes sense. Changed!

File: posts/2022-07-25-keyword-generics.md
Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md]: Agreed; added more clarifying comments and split it into two code blocks.

File: posts/2022-07-25-keyword-generics.md
Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md]: Changed the focus of this section. I agree it's probably not the right thing to highlight.

File: posts/2022-07-25-keyword-generics.md
Comment (oli-obk) [reply posts/2022-07-25-keyword-generics.md]: you can always inline a `const fn` at its call site. After a certain complexity that is not useful though. Showing a complex example here is not very helpful to the reader either. So I chose to write a simple `add_one` function, but honestly it could be anything that doesn't require iteration or branching (as that wasn't supported before `const fn`)

File: posts/2022-07-25-keyword-generics.md
Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md]: Good call; thank you!

File: posts/2022-07-25-keyword-generics.md
Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md]: Good call!

File: posts/2022-07-25-keyword-generics.md
Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md]: Incorporated your earlier feedback; now it does ^^

File: posts/2022-07-25-keyword-generics.md
Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md]: Yep, true. Removed!

**Comment by yoshuawuyts** [comment]
Updated with all feedback! - Also changed to date to tomorrow (Tuesday) to publish as it's getting quite late today.

File: posts/inside-rust/2022-07-26-keyword-generics.md
Comment (rylev) [inline posts/inside-rust/2022-07-26-keyword-generics.md:33]:
```suggestion
Rust allows you to be generic over types - it does not allow you to be generic over other things that are usually specified by keywords. For example, whether a function is async, whether a function can fail or not, whether a function is const or not, etc.
```

File: posts/inside-rust/2022-07-26-keyword-generics.md
Comment (rylev) [inline posts/inside-rust/2022-07-26-keyword-generics.md:63]: Would a footnote or `>` block look better?

File: posts/inside-rust/2022-07-26-keyword-generics.md
Comment (rylev) [inline posts/inside-rust/2022-07-26-keyword-generics.md:248]: Why two blocks?

**Review by rylev — APPROVED** [review]

File: posts/inside-rust/2022-07-26-keyword-generics.md
Comment (Mark-Simulacrum) [inline posts/inside-rust/2022-07-26-keyword-generics.md:284]:
```suggestion
exploring is not knowing what the outcomes will be. Right now we're in the
```

File: posts/inside-rust/2022-07-26-keyword-generics.md
Comment (yoshuawuyts) [reply posts/inside-rust/2022-07-26-keyword-generics.md:66]: Changed it to use `>`!

**Review by oli-obk — APPROVED** [review]

File: posts/2022-07-25-keyword-generics.md
Comment (yoshuawuyts) [reply posts/2022-07-25-keyword-generics.md]: Can we mark this as resolved then?

**Comment by oli-obk** [comment]
You need to mark it as `release: false`, that flag is about whether the post is a rust version release, not whether to publish the post.

I can press the merge button, I don't know if I'm allowed to XD who makes the final decision on blog posts?

**Comment by yoshuawuyts** [comment]
> You need to mark it as release: false, that flag is about whether the post is a rust version release, not whether to publish the post.

I don't believe such a flag is needed for the inside Rust blog? Here's the front matter from a previous post, which matches most other posts too:

```yaml
---
layout: post
title: Rust Compiler June 2022 Steering Cycle
author: Felix Klock
description: "The compiler team's June 2022 steering cycle"
team: The Compiler Team <https://www.rust-lang.org/governance/teams/compiler>
---
```

> I can press the merge button, I don't know if I'm allowed to XD who makes the final decision on blog posts?

My understanding from talking about this with the lang team is that any Rust team can publish to the inside rust blog, as long as it's relevant to the Rust project and we follow proper review etiquette - like we've done. I believe this is the reason why when you're added to a Rust initiative, you get commit access on the blog as well. Ideally this would all be documented in a `CONTRIBUTING.md` file on the repo, but there is none currently.

**Pull request merged**
